File: Cpp/Remove Duplicates from Sorted List II.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution{
public:
    /**
     * @param head: The first node of linked list.
     * @return: head node
     */
    ListNode * deleteDuplicates(ListNode *head) {
        // write your code here
        if (head == NULL) return head;
        map<int, int> st;
        ListNode *node = head;
        while (node != NULL) {
            if (st.find(node->val) != st.end()) st[node->val] ++;
            else st[node->val] = 1;
            node = node->next;
        }
        node = head;
        ListNode *pre = new ListNode(0);
        ListNode *res = pre;
        while (node != NULL) {
            if (st[node->val] == 1) {
                pre->next = node;
                pre = pre->next;
            }
            node = node->next;
        }
        pre->next = NULL;
        return res->next;
    }
};

File: Java/Remove-Duplicates-from-Sorted-List-II.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param ListNode head is the head of the linked list
     * @return: ListNode head of the linked list
     */
    public static ListNode deleteDuplicates(ListNode head) {
        // write your code here
        if (head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy;
        ListNode p = head;
        while (p != null) {
        	ListNode temp = p.next;
        	int num = 0;
        	while (temp != null && temp.val == p.val) {
        		num ++;
        		temp = temp.next;
        	}
        	if (num == 0) {
        		cur.next = p;
        		cur = cur.next;
        	}
        	p = temp;
        }
        cur.next = null;
        return dummy.next;
    }
}




